habitat_baselines/common/rollout_storage.py

```diff
@@ -4,237 +4,196 @@
-        self.actions = self.actions.to(device)
-        self.prev_actions = self.prev_actions.to(device)
-        self.masks = self.masks.to(device)
+        self.buffers.map_in_place(lambda v: v.to(device))
 
     def insert(
         self,
-        observations,
-        recurrent_hidden_states,
-        actions,
-        action_log_probs,
-        value_preds,
-        rewards,
-        masks,
+        observations=None,
+        recurrent_hidden_states=None,
+        actions=None,
+        action_log_probs=None,
+        value_preds=None,
+        rewards=None,
+        masks=None,
+        buffer_index: int = 0,
     ):
-        for sensor in observations:
-            self.observations[sensor][self.step + 1].copy_(
-                observations[sensor]
-            )
-        self.recurrent_hidden_states[self.step + 1].copy_(
-            recurrent_hidden_states
+        if not self.is_double_buffered:
+            assert buffer_index == 0
+
+        next_step = dict(
+            observations=observations,
+            recurrent_hidden_states=recurrent_hidden_states,
+            prev_actions=actions,
+            masks=masks,
         )
-        self.actions[self.step].copy_(actions)
-        self.prev_actions[self.step + 1].copy_(actions)
-        self.action_log_probs[self.step].copy_(action_log_probs)
-        self.value_preds[self.step].copy_(value_preds)
-        self.rewards[self.step].copy_(rewards)
-        self.masks[self.step + 1].copy_(masks)
 
-        self.step = self.step + 1
+        current_step = dict(
+            actions=actions,
+            action_log_probs=action_log_probs,
+            value_preds=value_preds,
+            rewards=rewards,
+        )
 
-    def after_update(self):
-        for sensor in self.observations:
-            self.observations[sensor][0].copy_(
-                self.observations[sensor][self.step]
-            )
+        next_step = {k: v for k, v in next_step.items() if v is not None}
+        current_step = {k: v for k, v in current_step.items() if v is not None}
 
-        self.recurrent_hidden_states[0].copy_(
-            self.recurrent_hidden_states[self.step]
+        env_slice = slice(
+            int(buffer_index * self._num_envs / self._nbuffers),
+            int((buffer_index + 1) * self._num_envs / self._nbuffers),
         )
-        self.masks[0].copy_(self.masks[self.step])
-        self.prev_actions[0].copy_(self.prev_actions[self.step])
-        self.step = 0
+
+        if len(next_step) > 0:
+            self.buffers.set(
+                (self.steps[buffer_index] + 1, env_slice),
+                next_step,
+                strict=False,
+            )
+
+        if len(current_step) > 0:
+            self.buffers.set(
+                (self.steps[buffer_index], env_slice),
+                current_step,
+                strict=False,
+            )
+
+    def advance_rollout(self, buffer_index: int = 0):
```

Comment:
Method is not used.

Reply:
It is used in ppo_trainer: https://github.com/facebookresearch/habitat-lab/blob/etw-updates/habitat_baselines/rl/ppo/ppo_trainer.py#L504